File: Game/Prototype.cs
namespace prototype.Game
{
    using System.Diagnostics;

    public class Prototype
    {
        private readonly Stopwatch timer_ = new Stopwatch();
        private readonly TextAnalyzer textAnalyzer_;
        private readonly MetricCalculator metricCalculator_;
        private readonly TextProcessor textProcessor_;
        private string fullText_;
        private string text_;

        public string Text
        {
            get => text_;
            set
            {
                fullText_ = value;
                text_ = value;
            }
        }

        public Prototype(
            TextAnalyzer textAnalyzer,
            MetricCalculator metricCalculator,
            TextProcessor textProcessor)
        {
            textAnalyzer_ = textAnalyzer;
            metricCalculator_ = metricCalculator;
            textProcessor_ = textProcessor;
        }

        public void LoadFile(string path)
        {
            Text = textProcessor_.LoadFile(path);
        }

        public bool Type(char key)
        {
            if (!timer_.IsRunning)
            {
                timer_.Start();
            }

            if (text_.Length > 0 && Text[0] != key)
            {
                return false;
            }

            text_ = text_.Substring(1);
            if (text_.Length == 0)
            {
                timer_.Stop();
            }

            return true;
        }

        public int WordsPerMinute()
        {
            var wordCount = textAnalyzer_.CountWords(fullText_);
            return metricCalculator_.WordsPerMinute(wordCount, timer_.ElapsedMilliseconds);
        }
    }
}

File: src/prototype.py
import textprocessing
import textanalysis
import metrics


class Prototype(object):
    @property
    def text(self):
        return self._text

    @text.setter
    def text(self, value):
        self._full_text = value
        self._text = value

    def load_file(self, path):
        self.text = textprocessing.load_file(path)

    def type_letter(self, letter):
        if len(self.text) > 0 and self.text[0] != letter:
            return False
        self._text = self._text[1:]
        if not self.text:
            pass
        return True

    def wpm(self):
        word_count = textanalysis.count_words(self._full_text)
        wpm = metrics.wpm(word_count, 60000)  # TODO
        return wpm
